Index: biod/BUILD.gn
===================================================================
--- biod.orig/BUILD.gn
+++ biod/BUILD.gn
@@ -7,8 +7,6 @@ import("//common-mk/pkg_config.gni")
 group("all") {
   deps = [
     ":bio_crypto_init",
-    ":bio_fw_updater",
-    ":bio_wash",
     ":biod",
     ":biod_client_tool",
     ":libbiod",
@@ -106,11 +104,11 @@ executable("bio_crypto_init") {
   deps = [ ":libbiod" ]
 }
 
-executable("bio_wash") {
-  sources = [ "tools/bio_wash.cc" ]
-  configs += [ ":target_defaults" ]
-  deps = [ ":libbiod" ]
-}
+#executable("bio_wash") {
+#  sources = [ "tools/bio_wash.cc" ]
+#  configs += [ ":target_defaults" ]
+#  deps = [ ":libbiod" ]
+#}
 
 executable("biod_client_tool") {
   sources = [ "tools/biod_client_tool.cc" ]
@@ -118,11 +116,11 @@ executable("biod_client_tool") {
   libs = [ "biod_proxy" ]
 }
 
-executable("bio_fw_updater") {
-  sources = [ "tools/bio_fw_updater.cc" ]
-  configs += [ ":target_defaults" ]
-  deps = [ ":libbiod" ]
-}
+#executable("bio_fw_updater") {
+#  sources = [ "tools/bio_fw_updater.cc" ]
+#  configs += [ ":target_defaults" ]
+#  deps = [ ":libbiod" ]
+#}
 
 if (use.test) {
   pkg_config("libchrome_test_config") {
