diff --git a/common-mk/platform2.py b/common-mk/platform2.py
index 6ea3c296ee..fb9727ce23 100755
--- a/common-mk/platform2.py
+++ b/common-mk/platform2.py
@@ -470,7 +470,7 @@ class Platform2(object):
                                     cwd=self.get_platform2_root(),
                                     stdout=True,
                                     encoding='utf-8')
-        return json.loads(result.output)
+        return json.loads(result.stdout)
 
     def compile(self, args):
         """Runs the compile step of the Platform2 build.
