Index: libfprint-v1.94.6/data/autosuspend.hwdb
===================================================================
--- libfprint-v1.94.6.orig/data/autosuspend.hwdb
+++ libfprint-v1.94.6/data/autosuspend.hwdb
@@ -5,6 +5,7 @@
 usb:v08FFp1600*
  ID_AUTOSUSPEND=1
  ID_PERSIST=0
+ ID_MISC_FINGERPRINT=1
 
 # Supported by libfprint driver aes1660
 usb:v08FFp1660*
@@ -26,18 +27,21 @@ usb:v08FFp168E*
 usb:v08FFp168F*
  ID_AUTOSUSPEND=1
  ID_PERSIST=0
+ ID_MISC_FINGERPRINT=1
 
 # Supported by libfprint driver aes2501
 usb:v08FFp2500*
 usb:v08FFp2580*
  ID_AUTOSUSPEND=1
  ID_PERSIST=0
+ ID_MISC_FINGERPRINT=1
 
 # Supported by libfprint driver aes2550
 usb:v08FFp2550*
 usb:v08FFp2810*
  ID_AUTOSUSPEND=1
  ID_PERSIST=0
+ ID_MISC_FINGERPRINT=1
 
 # Supported by libfprint driver aes2660
 usb:v08FFp2660*
@@ -60,22 +64,26 @@ usb:v08FFp268F*
 usb:v08FFp2691*
  ID_AUTOSUSPEND=1
  ID_PERSIST=0
+ ID_MISC_FINGERPRINT=1
 
 # Supported by libfprint driver aes3500
 usb:v08FFp5731*
  ID_AUTOSUSPEND=1
  ID_PERSIST=0
+ ID_MISC_FINGERPRINT=1
 
 # Supported by libfprint driver aes4000
 usb:v08FFp5501*
  ID_AUTOSUSPEND=1
  ID_PERSIST=0
+ ID_MISC_FINGERPRINT=1
 
 # Supported by libfprint driver egis0570
 usb:v1C7Ap0570*
 usb:v1C7Ap0571*
  ID_AUTOSUSPEND=1
  ID_PERSIST=0
+ ID_MISC_FINGERPRINT=1
 
 # Supported by libfprint driver elan
 usb:v04F3p0903*
@@ -141,6 +149,7 @@ usb:v04F3p0C6E*
 usb:v04F3p0C58*
  ID_AUTOSUSPEND=1
  ID_PERSIST=0
+ ID_MISC_FINGERPRINT=1
 
 # Supported by libfprint driver elanmoc
 usb:v04F3p0C7D*
@@ -152,11 +161,13 @@ usb:v04F3p0C8D*
 usb:v04F3p0C99*
  ID_AUTOSUSPEND=1
  ID_PERSIST=0
+ ID_MISC_FINGERPRINT=1
 
 # Supported by libfprint driver etes603
 usb:v1C7Ap0603*
  ID_AUTOSUSPEND=1
  ID_PERSIST=0
+ ID_MISC_FINGERPRINT=1
 
 # Supported by libfprint driver fpcmoc
 usb:v10A5pFFE0*
@@ -195,11 +206,13 @@ usb:v27C6p659C*
 usb:v27C6p6A94*
  ID_AUTOSUSPEND=1
  ID_PERSIST=0
+ ID_MISC_FINGERPRINT=1
 
 # Supported by libfprint driver nb1010
 usb:v298Dp1010*
  ID_AUTOSUSPEND=1
  ID_PERSIST=0
+ ID_MISC_FINGERPRINT=1
 
 # Supported by libfprint driver synaptics
 usb:v06CBp00BD*
@@ -218,6 +231,7 @@ usb:v06CBp015F*
 usb:v06CBp0104*
  ID_AUTOSUSPEND=1
  ID_PERSIST=0
+ ID_MISC_FINGERPRINT=1
 
 # Supported by libfprint driver upeksonly
 usb:v147Ep2016*
@@ -225,6 +239,7 @@ usb:v147Ep1000*
 usb:v147Ep1001*
  ID_AUTOSUSPEND=1
  ID_PERSIST=0
+ ID_MISC_FINGERPRINT=1
 
 # Supported by libfprint driver upektc
 usb:v0483p2015*
@@ -232,11 +247,13 @@ usb:v0483p2017*
 usb:v147Ep3001*
  ID_AUTOSUSPEND=1
  ID_PERSIST=0
+ ID_MISC_FINGERPRINT=1
 
 # Supported by libfprint driver upektc_img
 usb:v147Ep2020*
  ID_AUTOSUSPEND=1
  ID_PERSIST=0
+ ID_MISC_FINGERPRINT=1
 
 # Supported by libfprint driver uru4000
 usb:v045Ep00BC*
@@ -247,27 +264,32 @@ usb:v05BAp0008*
 usb:v05BAp000A*
  ID_AUTOSUSPEND=1
  ID_PERSIST=0
+ ID_MISC_FINGERPRINT=1
 
 # Supported by libfprint driver vcom5s
 usb:v061Ap0110*
  ID_AUTOSUSPEND=1
  ID_PERSIST=0
+ ID_MISC_FINGERPRINT=1
 
 # Supported by libfprint driver vfs0050
 usb:v138Ap0050*
  ID_AUTOSUSPEND=1
  ID_PERSIST=0
+ ID_MISC_FINGERPRINT=1
 
 # Supported by libfprint driver vfs101
 usb:v138Ap0001*
  ID_AUTOSUSPEND=1
  ID_PERSIST=0
+ ID_MISC_FINGERPRINT=1
 
 # Supported by libfprint driver vfs301
 usb:v138Ap0005*
 usb:v138Ap0008*
  ID_AUTOSUSPEND=1
  ID_PERSIST=0
+ ID_MISC_FINGERPRINT=1
 
 # Supported by libfprint driver vfs5011
 usb:v138Ap0010*
@@ -277,11 +299,13 @@ usb:v138Ap0017*
 usb:v138Ap0018*
  ID_AUTOSUSPEND=1
  ID_PERSIST=0
+ ID_MISC_FINGERPRINT=1
 
 # Supported by libfprint driver vfs7552
 usb:v138Ap0091*
  ID_AUTOSUSPEND=1
  ID_PERSIST=0
+ ID_MISC_FINGERPRINT=1
 
 # Known unsupported devices
 usb:v04E8p730B*
@@ -382,3 +406,8 @@ usb:v298Dp2033*
 usb:v3538p0930*
  ID_AUTOSUSPEND=1
  ID_PERSIST=0
+ ID_MISC_FINGERPRINT=1
+
+acpi:ELAN7001:
+acpi:ELAN70A1:
+ ID_MISC_FINGERPRINT=1
